vm/vm.c

```diff
@@ -0,0 +1,348 @@
+            else r0 = *(intptr_t*)str->str;
+            break;
+        case SI:
+            *(int64_t*)*sp++ = r0; // store int
+            break;
+        case SC:
+            r0 = *(char*)* sp++ = r0; // store char
+            break;
+        case PSH:
+            *--sp = r0; // push
+            break;
+        case OR:
+            r0 = *sp++ | r0;
+            break;
+        case XOR:
+            r0 = *sp++ ^ r0;
+            break;
+        case AND:
+            r0 = *sp++ & r0;
+            break;
+        case EQ:
+            r0 = *sp++ == r0;
+            break;
+        case NE:
+            r0 = *sp++ != r0;
+            break;
+        case LT:
+            r0 = *sp++ < r0;
+            break;
+        case GT:
+            r0 = *sp++ > r0;
+            break;
+        case LE:
+            r0 = *sp++ <= r0;
+            break;
+        case GE:
+            r0 = *sp++ >= r0;
+            break;
+        case SHL:
+            r0 = *sp++ << r0;
+            break;
+        case SHR:
+            r0 = *sp++ >> r0;
+            break;
+        case ADD:
+            r0 = *sp++ + r0;
+            break;
+        case SUB:
+            r0 = *sp++ - r0;
+            break;
+        case MUL:
+            r0 = *sp++ * r0;
+            break;
+        case DIV:
+            r0 = *sp++ / r0;
+            break;
+        case MOD:
+            r0 = *sp++ % r0;
+            break;
+        case OPEN:
+            r0 = open((char*)sp[1], *sp);
+            break;
+        case READ:
+            r0 = read(sp[2], (char*)sp[1], *sp);
+            break;
+        case WRIT:
+            r0 = write(sp[2], (char*)sp[1], *sp);
+            break;
+        case CLOS:
+            r0 = close(*sp);
+            break;
+        case PRTF:
+            assert((pc + 1)->ir == ADJ);
+            nu = (pc + 1)->arg;
+            args = malloc(sizeof(intptr_t) * nu);
+            fetch_args(sp, args, nu);
+            char pr_buff[MAX_BUF];
+            fstring(pr_buff, MAX_BUF, (char*)args[0], args + 1 , nu - 1);
+            r0 = printf("%s", pr_buff);
+            free(args);
+            break;
+        case MALC:
+            r0 = (int64_t)malloc(*sp);
+            break;
+        case FREE:
+            break;
+        case MSET:
+            r0 = (int64_t)memset((char*)sp[2], sp[1], *sp);
+            break;
+        case MCMP:
+            r0 = memcmp((char*)sp[2], (char*)sp[1], *sp);
+            break;
+        case MCPY:
+            r0 = (int64_t)memcpy((char*)sp[2], (char*)sp[1], *sp);
+            break;
+        case MMAP:
+            r0 = mmap((char*)sp[5], (char*)sp[4], (char*)sp[3], (char*)sp[2], (char*)sp[1], *sp);
+            break;
+        case DSYM:
+            r0 = dlsym(sp[1], *sp);
+            break;
+        case DLOP:
+            r0 = dlopen((char*)sp[1], *sp);
+            break;
+        case BSCH:
+            r0 = bsearch(sp[4], sp[3], sp[2], sp[1], *sp);
+            break;
+        case STRL:
```

Comment:
How about eliminating `strlen` as an exposed function? That is, we can implement our own string length routine without declaring.

Reply:
It can be `_strlen` or something alike.

Reply:
> How about eliminating `strlen` as an exposed function? That is, we can implement our own string length routine without declaring.

Yes, we can. But is will generate more code length.  Change it?

Reply:
Implementing strlen only takes some lines of code.